src/HeaderFields.cpp

```diff
@@ -0,0 +1,125 @@
+#include "HeaderFields.hpp"
+
+#include <algorithm>
+#include <cctype>
+#include <cstddef>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <vector>
+
+std::string HeaderFields::toLower(std::string s)
+{
+	std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c)
+		{ return std::tolower(c); });
+	return s;
+}
+
+std::string HeaderFields::trim(const std::string &s)
+{
+	size_t start = s.find_first_not_of(" \t\r\n");
+	if (start == std::string::npos)
+		return "";
+	size_t end = s.find_last_not_of(" \t\r\n");
+	return s.substr(start, end - start + 1);
+}
+
+void HeaderFields::parse(std::string_view data)
+{
+	size_t pos = 0;
+	while (pos < data.size())
+	{
+		// Find end of line
+		size_t lineEnd = data.find("\r\n", pos);
+		bool hasCRLF = (lineEnd != std::string_view::npos);
+		if (!hasCRLF)
+			lineEnd = data.size();
+
+		std::string_view line = data.substr(pos, lineEnd - pos);
+		pos = hasCRLF ? lineEnd + 2 : lineEnd; // Skip \r\n only if present
+
+		// Skip empty lines
+		if (line.empty())
+			continue;
+
+		// Find colon separator
+		size_t colonPos = line.find(':');
+		if (colonPos == std::string_view::npos)
+			throw std::runtime_error("malformed header: no colon separator");
+
+		std::string key = trim(std::string(line.substr(0, colonPos)));
+		std::string value = trim(std::string(line.substr(colonPos + 1)));
+
+		if (key.empty())
+			throw std::runtime_error("malformed header: empty key");
+
+		// Store with lowercase key
+		headers_[toLower(key)].push_back(value);
+	}
+}
+
+std::string HeaderFields::get(const std::string &key) const
+{
+	auto it = headers_.find(toLower(key));
+	if (it == headers_.end())
+		return std::string();
+
+	const std::vector<std::string> &values = it->second;
+	std::string result;
+	for (size_t i = 0; i < values.size(); ++i)
+	{
+		if (i > 0)
+			result += ", ";
+		result += values[i];
+	}
+	return result;
+}
+
+bool HeaderFields::has(const std::string &key) const
+{
+	return headers_.find(toLower(key)) != headers_.end();
+}
+
+bool HeaderFields::remove(const std::string &key)
+{
+	auto it = headers_.find(toLower(key));
+	if (it == headers_.end())
+		return false;
+	headers_.erase(it);
+	return true;
+}
+
+void HeaderFields::set(const std::string &key, const std::string &value)
+{
+	std::string lowerKey = toLower(key);
+	headers_[lowerKey].clear();
+	headers_[lowerKey].push_back(value);
+}
+
+void HeaderFields::add(const std::string &key, const std::string &value)
+{
+	headers_[toLower(key)].push_back(value);
+}
+
+const std::map<std::string, std::vector<std::string>> &HeaderFields::all() const
+{
+	return headers_;
+}
+
+std::string HeaderFields::serialize() const
+{
+	std::string result;
+	for (const auto &pair : headers_)
+	{
+		result += pair.first + ": ";
+		for (size_t i = 0; i < pair.second.size(); ++i)
+		{
+			if (i > 0)
+				result += ", ";
+			result += pair.second[i];
+		}
```

Comment:
https://www.rfc-editor.org/rfc/rfc9110#name-field-order points out that Set-Cookie must be special cased to not combine.

Reply:
Good call out! We've separated special case logic for set-cookie to be handled as required.